src/services/prices/price-sources/codex-price-source.ts

```diff
@@ -0,0 +1,182 @@
+      getCurrentPrices: true,
+      getHistoricalPrices: true,
+      getBulkHistoricalPrices: true,
+      getChart: false,
+    };
+    const entries = SUPPORTED_CHAINS.map(({ chainId }) => chainId).map((chainId) => [chainId, support]);
+    return Object.fromEntries(entries);
+  }
+
+  async getCurrentPrices({
+    tokens,
+    config,
+  }: {
+    tokens: PriceInput[];
+    config: { timeout?: TimeString } | undefined;
+  }): Promise<Record<ChainId, Record<TokenAddress, PriceResult>>> {
+    const input = tokens.map(({ token, chainId }) => ({ chainId, token }));
+    const prices = await this.getBulkPrices({ tokens: input, config });
+    return Object.fromEntries(
+      Object.entries(prices).map(([chainId, tokens]) => [
+        chainId,
+        Object.fromEntries(Object.entries(tokens).map(([token, price]) => [token, Object.values(price).at(0)!])),
+      ])
+    );
+  }
+
+  async getHistoricalPrices({
+    tokens,
+    timestamp,
+    searchWidth,
+    config,
+  }: {
+    tokens: PriceInput[];
+    timestamp: Timestamp;
+    searchWidth: TimeString | undefined;
+    config: { timeout?: TimeString } | undefined;
+  }): Promise<Record<ChainId, Record<TokenAddress, PriceResult>>> {
+    const input = tokens.map(({ token, chainId }) => ({ chainId, token, timestamp }));
+    const prices = await this.getBulkHistoricalPrices({ tokens: input, searchWidth, config });
+    return Object.fromEntries(
+      Object.entries(prices).map(([chainId, tokens]) => [
+        chainId,
+        Object.fromEntries(Object.entries(tokens).map(([token, price]) => [token, price[timestamp]])),
+      ])
+    );
+  }
+
+  getBulkHistoricalPrices({
+    tokens,
+    searchWidth,
+    config,
+  }: {
+    tokens: { chainId: ChainId; token: TokenAddress; timestamp: Timestamp }[];
+    searchWidth: TimeString | undefined;
+    config: { timeout?: TimeString } | undefined;
+  }): Promise<Record<ChainId, Record<TokenAddress, Record<Timestamp, PriceResult>>>> {
+    return this.getBulkPrices({ tokens, searchWidth, config });
+  }
+
+  async getChart(_: {
+    tokens: PriceInput[];
+    span: number;
+    period: TimeString;
+    bound: { from: Timestamp } | { upTo: Timestamp | 'now' };
+    searchWidth?: TimeString;
+    config: { timeout?: TimeString } | undefined;
+  }): Promise<Record<ChainId, Record<TokenAddress, PriceResult[]>>> {
+    return Promise.reject(new Error('Operation not supported'));
+  }
+
+  private async getBulkPrices({
+    tokens,
+    searchWidth,
+    config,
+  }: {
+    tokens: { chainId: ChainId; token: TokenAddress; timestamp?: Timestamp }[];
+    searchWidth?: TimeString;
+    config: { timeout?: TimeString } | undefined;
+  }): Promise<Record<ChainId, Record<TokenAddress, Record<Timestamp, PriceResult>>>> {
+    const result: Record<ChainId, Record<TokenAddress, Record<Timestamp, PriceResult>>> = {};
+    if (!this.apiKey) {
+      return result;
+    }
+    const chunks = splitInChunks(tokens, 25);
```

Comment:
Is this necessary? We did this with DefiLlama because the params were passed query params, but since this is a POST, maybe it isn't necessary
Did you find any limitations that forced you to add this?

Reply:
Yes, the docs say it support up to 25 tokens per request